tests/resolver/test_lib.py

```diff
@@ -0,0 +1,389 @@
+from __future__ import annotations
+
+import pytest
+
+from variantlib.errors import ValidationError
+from variantlib.models.variant import VariantDescription
+from variantlib.models.variant import VariantFeature
+from variantlib.models.variant import VariantProperty
+from variantlib.resolver.filtering import filter_variants_by_features
+from variantlib.resolver.filtering import filter_variants_by_namespaces
+from variantlib.resolver.filtering import remove_duplicates
+from variantlib.resolver.lib import filter_variants
+from variantlib.resolver.lib import sort_and_filter_supported_variants
+
+
+@pytest.fixture(scope="session")
+def vprops() -> list[VariantProperty]:
+    """Fixture to create a list of VariantProperty objects."""
+
+    return [
+        VariantProperty(namespace="OmniCorp", feature="access_key", value="value1"),
+        VariantProperty(
+            namespace="TyrellCorporation", feature="client_id", value="value2"
+        ),
+        VariantProperty(
+            namespace="TyrellCorporation", feature="client_pass", value="abcde"
+        ),
+        VariantProperty(
+            namespace="TyrellCorporation", feature="client_pass", value="efghij"
+        ),
+    ]
+
+
+@pytest.fixture(scope="session")
+def vdescs(vprops: list[VariantProperty]) -> list[VariantDescription]:
+    """Fixture to create a list of VariantDescription objects."""
+
+    assert len(vprops) == 4
+    prop1, prop2, prop3, prop4 = vprops
+
+    return [
+        # prop3 and prop4 are mutually exclusive
+        VariantDescription([prop1, prop2, prop3]),
+        VariantDescription([prop1, prop3, prop2]),
+        VariantDescription([prop2, prop3, prop1]),
+        VariantDescription([prop2, prop1, prop3]),
+        VariantDescription([prop3, prop2, prop1]),
+        VariantDescription([prop3, prop1, prop2]),
+        VariantDescription([prop1, prop2, prop4]),  # duplicate with prop4
+        VariantDescription([prop1, prop4, prop2]),  # duplicate with prop4
+        VariantDescription([prop2, prop4, prop1]),  # duplicate with prop4
+        VariantDescription([prop2, prop1, prop4]),  # duplicate with prop4
+        VariantDescription([prop4, prop2, prop1]),  # duplicate with prop4
+        VariantDescription([prop4, prop1, prop2]),  # duplicate with prop4
+        VariantDescription([prop1, prop2]),
+        VariantDescription([prop2, prop1]),
+        VariantDescription([prop1, prop3]),
+        VariantDescription([prop3, prop1]),
+        VariantDescription([prop2, prop3]),
+        VariantDescription([prop3, prop2]),
+        VariantDescription([prop1, prop4]),  # duplicate with prop4
+        VariantDescription([prop4, prop1]),  # duplicate with prop4
+        VariantDescription([prop2, prop4]),  # duplicate with prop4
+        VariantDescription([prop4, prop2]),  # duplicate with prop4
+        VariantDescription([prop1]),
+        VariantDescription([prop2]),
+        VariantDescription([prop3]),
+        VariantDescription([prop4]),
+    ]
```

Comment:
```suggestion
    return list(y for y in itertools.chain.from_iterable(itertools.permutations(vprops, x+1) for x in range(len(vprops))) if not (prop3 in y and prop4 in y))
```

EDIT: I've just realized you're depending on specific order here, while `itertools.permutations()` gives a different (but stable) ordering.